### PR TITLE
UI Plugin System

### PR DESCRIPTION
**Pull request by msavela**
Two ways to register a new plugin to UI service:

1. Placing a folder under `services/ui_backend_service/plugins/installed`
2. Defining a remote Git repository location

Plugins can be registered via `PLUGINS` environment variable. The value should be a _stringified_ json of the format:

```json
{
  "plugin-example": "https://github.com/Netflix/metaflow-ui-plugin-example.git"
}
```

All plugins are installed under `services/ui_backend_service/plugins/installed`. Local plugins should be placed under this folder, e.g. `services/ui_backend_service/plugins/installed/plugin-example/`.
Plugins are loaded only if `PLUGINS` environment variable contains entry for a specific plugin.

## Documentation

At minimun a plugin should contain file called `manifest.json` with following contents:

```json
{
  "name": "plugin-example",
  "version": "1.0.0",
  "entrypoint": "index.html"
}
```

Where `entrypoint` refers to a source that will be loaded inside sandboxed iframe on user's browser.

`entrypoint` can also be defined as absolute url:

```json
{
  ...
  "entrypoint": "https://hostname/index.html"
}
```

Following API routes are supported:

| Route                                  | Description                                                                   |
| -------------------------------------- | ----------------------------------------------------------------------------- |
| `GET /plugin`                          | List all successfully registered plugins                                      |
| `GET /plugin/{plugin_name}`            | List plugin details                                                           |
| `GET /plugin/{plugin_name}/{filename}` | Serve plugin files, such as `manifest.json`, `README.md` or `dist/index.html` |

## Plugin development

**TODO:** Following links are not valid yet.

- Plugin development documentation can be found from [Netflix/metaflow-ui](https://github.com/Netflix/metaflow-ui) repository.
- See [example plugin](https://github.com/Netflix/metaflow-ui-plugin-example) as a reference implementation.

## Authentication credentials

There are multiple ways to provide authentication credentials:

- Username and password
- SSH key pair
- SSH Agent (`~/.ssh`)

Authentication credentials can be provided by using `auth` object at the top level of `PLUGINS` json
or alternatively by defining `auth` object at repository level.

```json
{
  "auth": {
    "public_key": "/path/to/id_rsa.pub",
    "private_key": "/path/to/id_rsa"
  },
  "metaflow-ui-plugin-example": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
  "metaflow-ui-plugin-noauth": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example-noauth.git",
    "auth": null
  }
}
```

## Examples

Following JSON describes different ways to register plugins. Each plugin will be automatically downloaded to `services/ui_backend_service/plugins/installed`:

```json
{
  "plugin": "https://github.com/Netflix/metaflow-ui-plugin-example.git",
  "plugin-repository": {
    "repository": "https://github.com/Netflix/metaflow-ui-plugin-example.git"
  },
  "plugin-branch": {
    "repository": "https://github.com/Netflix/metaflow-ui-plugin-example.git",
    "ref": "origin/feature/test"
  },
  "plugin-commit": {
    "repository": "https://github.com/Netflix/metaflow-ui-plugin-example.git",
    "ref": "3758f6a"
  },
  "plugin-commit-with-parameters": {
    "repository": "https://github.com/Netflix/metaflow-ui-plugin-example.git",
    "ref": "3758f6a",
    "parameters": {
      "color": "blue"
    }
  },
  "plugin-repository-multiple": {
    "repository": "https://github.com/Netflix/metaflow-ui-plugin-example.git",
    "paths": ["path/to/first-plugin", "second-plugin"]
  }
}
```

Following JSON describes different ways to provide authentication credentials for remote Git repositories:

```json
{
  "auth": {
    "public_key": "/root/id_rsa.pub",
    "private_key": "/root/id_rsa",
    "user": "git",
    "pass": "optional-passphrase"
  },
  "plugin-auth-global": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
  "plugin-noauth": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": null
  },
  "plugin-auth-user": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "user": "username"
    }
  },
  "plugin-auth-userpass": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "user": "username",
      "pass": "password"
    }
  },
  "plugin-auth-key-user": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "public_key": "ssh-rsa AAAA...",
      "private_key": "-----BEGIN RSA PRIVATE KEY-----...",
      "user": "custom"
    }
  },
  "plugin-auth-key-pass": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "public_key": "/root/id_rsa.pub",
      "private_key": "/root/id_rsa",
      "pass": "optional-passphrase"
    }
  },
  "plugin-auth-agent": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "agent": true
    }
  },
  "plugin-auth-agent-user": {
    "repository": "git@github.com:Netflix/metaflow-ui-plugin-example.git",
    "auth": {
      "agent": true,
      "user": "custom"
    }
  }
}
```

Following JSON describes different ways to register local plugins. Each plugin should already have folder under `services/ui_backend_service/plugins/installed` with the same name:

```json
{
  "plugin-local": {},
  "plugin-local-with-parameteres": {
    "parameters": {
      "color": "yellow"
    }
  },
  "plugin-local-multiple": {
    "paths": ["path/to/first-plugin", "second-plugin"]
  }
}
```

### CODE REVIEW

File: services/ui_backend_service/api/plugins.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        except:
+            return web_response(500, "Internal server error")
+
+
+def _get_plugin_from_request_old(request):
+    _plugins = list_plugins()
+    plugin_name = request.match_info.get("plugin_name")
+    if plugin_name not in _plugins:
+        return None
+    return _plugins[plugin_name]
```
Comment (saikonen) [inline services/ui_backend_service/api/plugins.py:103]: some forgotten test stuff?

File: services/ui_backend_service/api/plugins.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def _get_plugin_from_request(request):
+    _plugins = list_plugins()
+    plugin_name = request.match_info.get("plugin_name")
```
Comment (saikonen) [inline services/ui_backend_service/api/plugins.py:108]: not a blocker, but is there any check in place when setting up plugins for rejecting plugins with duplicate names?

File: services/ui_backend_service/api/plugins.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+def _plugin_dict(plugin):
+    return {
+        "identifier": plugin.identifier,
+        "name": plugin.name,
+        "repository": plugin.repository,
+        "ref": plugin.ref,
+        "parameters": plugin.parameters,
+        "config": plugin.config,
+        "files": plugin.files
+    }
```
Comment (saikonen) [inline services/ui_backend_service/api/plugins.py:124]: is this unique to the plugin api side, or should it be a getter (or magic method if possible, so you could do `dict(plugin)` instead) on the plugin object itself?

**Review by saikonen — APPROVED** [review]

File: services/ui_backend_service/api/plugins.py (hunk 2 above)
Comment (msavela) [reply services/ui_backend_service/api/plugins.py:108]: Python's `json.load` doesn't allow duplicate object keys in the first place and only returns the last key present in the object.

**Review by saikonen — APPROVED** [review]

**Pull request merged**